src/main/java/org/broadinstitute/hellbender/utils/mcmc/AbstractParameterizedState.java

```diff
@@ -0,0 +1,127 @@
+
+import java.util.*;
+
+/**
+ * Represents a collection of named {@link Parameter} objects, which may hold values of mixed type.
+ * See GibbsSamplerSingleGaussianUnitTest and GibbsSamplerCopyRatioUnitTest for examples of use.
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public abstract class AbstractParameterizedState {
+    private final Map<String, Parameter<?>> parameterMap = new HashMap<>();
+
+    /**
+     * Constructs an AbstractParameterizedState from a List of {@alink Parameter} objects with values of mixed type.
+     * @param parameters    List of Parameters
+     * @throws IllegalArgumentException if {@code parameters} is null or contains duplicate parameter names
+     */
+    public AbstractParameterizedState(final List<Parameter<?>> parameters) {
+        Utils.nonNull(parameters, "List of parameters cannot be null.");
+        for (final Parameter<?> parameter : parameters) {
+            if (parameterMap.containsKey(parameter.name())) {
+                throw new IllegalArgumentException("List of parameters cannot contain duplicate parameter names.");
+            }
+            parameterMap.put(parameter.name(), parameter);
+        }
+    }
+
+    /**
+     * Copy constructor for AbstractParameterizedState.
+     * @param state AbstractParameterizedState to copy
+     */
+    public AbstractParameterizedState(final AbstractParameterizedState state) {
+        this(new ArrayList<>(state.parameterMap.values()));
+    }
+
+    /**
+     * Constructs an AbstractParameterizedState consisting of a collection of {@link Parameter} objects of identical
+     * type, given a List of parameter values.  The parameter names are constructed by appending a numerical index
+     * to a common parameter-name prefix.
+     * @param parameterNamePrefix   common Parameter-name prefix
+     * @param parameterValues       List of Parameter values
+     * @param <T>                   common type for all Parameter values
+     * @see     AbstractParameterizedState#initializeParameters(String, List)
+     */
+    public <T> AbstractParameterizedState(final String parameterNamePrefix, final List<T> parameterValues) {
+        this(initializeParameters(parameterNamePrefix, parameterValues));
+    }
+
+    /**
+     * Returns the value of a {@link Parameter} contained in the collection held by the AbstractParameterizedState,
+     * given the parameter name and type.
+     * @param parameterName         name of Parameter
+     * @param parameterValueClass   class of Parameter value
+     * @param <T>                   type of Parameter value
+     * @return                      Parameter value
+     * @throws IllegalArgumentException if {@code parameterName} does not correspond to a Parameter in the collection or
+     *                                  if {@code parameterValueClass} does not match that of the Parameter in the collection
+     */
+    public <T> T get(final String parameterName, final Class<T> parameterValueClass) {
```

Comment:
As a check of my understanding: the ugly but concise way of doing this would be to have this return `Object`, thereby obviating the second argument but forcing you to use a bunch of casts downstream.  A compromise would be to basically wrap `Object` in a `ParameterValue` class, but you opted against this because it wouldn't be type-safe regarding different types of parameter values.  Do I grok?